includes/class-wc-admin-api-init.php

```diff
@@ -355,49 +429,146 @@ public static function customers_report_data_store_init() {
+	 * Returns the batch size for regenerating reports.
+	 * Note: can differ per batch action.
 	 *
-	 * @param WC_Background_Updater|null $updater Updater instance.
-	 * @return bool
+	 * @param string $action Single batch action name.
+	 * @return int Batch size.
 	 */
-	public static function customer_lookup_store_init( $updater = null ) {
-		// TODO: this needs to be updated a bit, as it no longer runs as a part of WC_Install, there is no bg updater.
-		global $wpdb;
+	public static function get_batch_size( $action ) {
+		$batch_sizes = array(
+			self::QUEUE_BATCH_ACTION     => 100,
+			self::CUSTOMERS_BATCH_ACTION => 25,
+			self::ORDERS_BATCH_ACTION    => 10,
+		);
+		$batch_size  = isset( $batch_sizes[ $action ] ) ? $batch_sizes[ $action ] : 25;
+
+		/**
+		 * Filter the batch size for regenerating a report table.
+		 *
+		 * @param int    $batch_size Batch size.
+		 * @param string $action Batch action name.
+		 */
+		return apply_filters( 'wc_admin_report_regenerate_batch_size', $batch_size, $action );
+	}
+
+	/**
+	 * Queue a large number of batch jobs, respecting the batch size limit.
+	 * Reduces a range of batches down to "single batch" jobs.
+	 *
+	 * @param int    $range_start Starting batch number.
+	 * @param int    $range_end Ending batch number.
+	 * @param string $single_batch_action Action to schedule for a single batch.
+	 * @return void
+	 */
+	public static function queue_batches( $range_start, $range_end, $single_batch_action ) {
+		$batch_size = self::get_batch_size( self::QUEUE_BATCH_ACTION );
+		$range_size = 1 + ( $range_end - $range_start );
+		$schedule   = time() + 5;
+
+		if ( $range_size > $batch_size ) {
+			// If the current batch range is larger than a single batch,
+			// split the range into $queue_batch_size chunks.
+			$chunk_size = ceil( $range_size / $batch_size );
+
+			for ( $i = 0; $i < $batch_size; $i++ ) {
+				$batch_start = $range_start + ( $i * $chunk_size );
+				$batch_end   = min( $range_end, $range_start + ( $chunk_size * ( $i + 1 ) ) - 1 );
+
+				self::queue()->schedule_single(
+					$schedule,
+					self::QUEUE_BATCH_ACTION,
+					array( $batch_start, $batch_end, $single_batch_action )
+				);
+			}
+		} else {
+			// Otherwise, queue the single batches.
+			for ( $i = $range_start; $i <= $range_end; $i++ ) {
+				self::queue()->schedule_single( $schedule, $single_batch_action, array( $i ) );
+			}
+		}
+	}
+
+	/**
+	 * Queue an action to run after another.
+	 *
+	 * @param string $action Action to run after prerequisite.
+	 * @param string $prerequisite_action Prerequisite action.
+	 */
+	public static function queue_dependent_action( $action, $prerequisite_action ) {
+		$blocking_jobs = self::queue()->search(
+			array(
+				'status'   => 'pending',
+				'orderby'  => 'date',
+				'order'    => 'DESC',
+				'per_page' => 1,
+				'claimed'  => false,
+				'search'   => $prerequisite_action, // search is used instead of hook to find queued batch creation.
+			)
+		);
```

Comment:
I'm not sure, but it might be worth checking that action does not stay in `pending` status if one of the prerequisites fail in some unexpected way. Otherwise the dependents would never run.

Unfortunately, I'm not sure if there is a better way... Maybe there could be a cleanup task scheduled, i.e. if the number of pending prerequisite jobs does no decrease in a day, unschedule all of them and reschedule again?

Reply:
Can you elaborate on this or perhaps say it in a different way? I'm not sure I'm following.

If a job throws an uncaught Exception, Action Scheduler will mark it as `failed` ([#](https://github.com/Prospress/action-scheduler/blob/7995d4232b3e4655669d1191c76e47fbce173e05/classes/ActionScheduler_Abstract_QueueRunner.php#L63)) - if the job just doesn't succeed, but doesn't throw any uncaught Exceptions, Action Scheduler will mark it as `complete`.

In both cases, that job would not be in the search results for `pending` actions.